### ISSUE TITLE
Mouse input not working with Civilization VI

### ISSUE THREAD

**Issue by acidicoala** [acidicoala/ScreamAPI#3]
Just keeping it here as a reminder that for some reason mouse input is not working on the achievement overlay, at least on DX11. Have not tested it with DX12 yet, need to implement multi-renderer support first. Will have to investigate this matter further later on.

**Comment by Nemirtingas** [comment]
Take a look at: https://gitlab.com/Mr_Goldberg/goldberg_emulator/-/commit/6d96784af178ac1893002bbfad53f6542bca8dd1

**Comment by acidicoala** [comment]
> Take a look at: https://gitlab.com/Mr_Goldberg/goldberg_emulator/-/commit/6d96784af178ac1893002bbfad53f6542bca8dd1

Will do later this evening

**Comment by acidicoala** [comment]
> Take a look at: https://gitlab.com/Mr_Goldberg/goldberg_emulator/-/commit/6d96784af178ac1893002bbfad53f6542bca8dd1

Unfortunately it didn't help either.

**Comment by Nemirtingas** [comment]
Then I guess you should add a debug line with the Mouse coordinates with ImGui's internal Mouse coordinates if you didn't already do it.

**Comment by acidicoala** [comment]
Cursor movement actually works fine, as well as hover events. Only the clicking/scrolling doesn't work...

**Comment by Nemirtingas** [comment]
Are you sure mouse really works like ImGui wants it ?
I see you have a fix:
```
POINT mPos;
GetCursorPos(&mPos);
ScreenToClient(hWnd, &mPos);
ImGui::GetIO().MousePos.x = (float)mPos.x;
ImGui::GetIO().MousePos.y = (float)mPos.y;
```

Maybe your mouse isn't detected by ImGui but you think it works because of that.

**Comment by acidicoala** [comment]
> POINT mPos;
> GetCursorPos(&mPos);
> ScreenToClient(hWnd, &mPos);
> ImGui::GetIO().MousePos.x = (float)mPos.x;
> ImGui::GetIO().MousePos.y = (float)mPos.y;

That was a mouse input fix for Remnant back when I was just starting to hook ImGui. The mouse was really messed up and this fixed it. However funny enough, now it works perfectly fine without it in Remnant, so I'm going to remove that bit of code. But unfortunately it didn't affect Civ VI in any way

**Comment by Nemirtingas** [comment]
The funky mouse could happen when a game doesn't use the mouse coordinates but use its own way to compute the mouse pos, like on Truck Simulator 2, they draw their own Mouse based on the direct input mouse event... So the ImGui, System and Game Cursor were all 3 at different positions, it could be the case on Remnant when playing windowed/fullscreen ?

I'll download the sources and see whats happening on Civ, maybe I can find something. That weird that the mouse click isn't seen if the keyboard events are ok.

**Comment by acidicoala** [comment]
Tbh I haven't tested keyboard events. But I know mouse hover events are working fine, you can see buttons or scroll bar being kind of focused when you hover mouse over them.

**Comment by Nemirtingas** [comment]
keyboard event should work, or you wouldn't be able to open the overlay ;).

**Comment by acidicoala** [comment]
Ah, of course. I'm being stupid today :)

**Comment by acidicoala** [comment]
Ok, I was able to kind of fix it. Apparently game (or whoever) sends different message codes when the overlay is open. More specifically, it sends messages from this range:
>#if(WINVER >= 0x0602)
#define WM_NCPOINTERUPDATE              0x0241
#define WM_NCPOINTERDOWN                0x0242
#define WM_NCPOINTERUP                  0x0243
#define WM_POINTERUPDATE                0x0245
#define WM_POINTERDOWN                  0x0246
#define WM_POINTERUP                    0x0247
#define WM_POINTERENTER                 0x0249
#define WM_POINTERLEAVE                 0x024A
#define WM_POINTERACTIVATE              0x024B
#define WM_POINTERCAPTURECHANGED        0x024C
#define WM_TOUCHHITTESTING              0x024D
#define WM_POINTERWHEEL                 0x024E
#define WM_POINTERHWHEEL                0x024F
#define DM_POINTERHITTEST               0x0250
#define WM_POINTERROUTEDTO              0x0251
#define WM_POINTERROUTEDAWAY            0x0252
#define WM_POINTERROUTEDRELEASED        0x0253
#endif /* WINVER >= 0x0602 */

For now I have made a dirty fix which sends the ImGui message code that it normally expects when one of those above is received:
```c++
if(showAchievementManager) {
	switch(uMsg)
	{
		case WM_POINTERDOWN:
			uMsg = WM_LBUTTONDOWN;
			break;
		case WM_POINTERUP:
			uMsg = WM_LBUTTONUP;
			break;
		case WM_POINTERWHEEL:
			uMsg = WM_MOUSEWHEEL;
			break;
		case WM_POINTERUPDATE:
			uMsg = WM_SETCURSOR;
			break;
	}
	ImGui_ImplWin32_WndProcHandler(hWnd, uMsg, wParam, lParam);
	return true;
} else {
	return CallWindowProc(originalWndProc, hWnd, uMsg, wParam, lParam);
}
```

If there is no better solution, I will keep it permanent.

**Comment by Nemirtingas** [comment]
Ask a fix in ImGui's repo then ;).

**Comment by acidicoala** [comment]
Kinda fixed in 32f87b2
